### ISSUE TITLE
LoadBalancer rules applied to NSG does not obey service tag configuration

### ISSUE THREAD

**Issue by nmiodice** [Azure/AKS#1813]
**What happened**:

When applying a `LoadBalancer` configuration, I need to lock down traffic to a particular service tag (`AzureFrontDoor.Backend`). When manually modifying the NSG in the portal I am able to get this to work. The IP is not accessible, unless through Front Door.

When following these [instructions](https://docs.microsoft.com/en-us/azure/aks/load-balancer-standard), I do see the NSG ingress rule re-created. It looks the same as the one I create in the portal. However, access via FrontDoor is **blocked**.

Is there anything on the cluster, perhaps with NGINX, that I need to be aware of?

Load Balancer configuration (without NSG rules). Applying this, then manually adding the NSG rule results in expected behavior.
```yaml
apiVersion: v1
kind: Service
metadata:
  labels:
    app: nginx-ingress
    chart: nginx-ingress-1.40.3
    component: controller
    heritage: Helm
    release: nginx-ingress
  name: nginx-ingress-controller
  namespace: nginx-ingress
spec:
  ports:
  - name: http
    port: 80
    protocol: TCP
    targetPort: http
  selector:
    app: nginx-ingress
    app.kubernetes.io/component: controller
    release: nginx-ingress
  type: LoadBalancer
```

Load Balancer configuration (WITH NSG rules)
```yaml
apiVersion: v1
kind: Service
metadata:
  annotations:
    service.beta.kubernetes.io/azure-allowed-service-tags: AzureFrontDoor.Backend
  labels:
    app: nginx-ingress
    chart: nginx-ingress-1.40.3
    component: controller
    heritage: Helm
    release: nginx-ingress
  name: nginx-ingress-controller
  namespace: nginx-ingress
spec:
  # without this, an ingress rule to allow 0.0.0.0/0 is created, which is not ideal!
  loadBalancerSourceRanges:
  - 127.0.0.1/24
  ports:
  - name: http
    port: 80
    protocol: TCP
    targetPort: http
  selector:
    app: nginx-ingress
    app.kubernetes.io/component: controller
    release: nginx-ingress
  type: LoadBalancer
```

**What you expected to happen**:

Applying the second YAML configuration that I provided should result in access from Front Door. 

**How to reproduce it (as minimally and precisely as possible)**:
Apply first YAML and then create an ingress rule for AFD. It should work. But the second one does not.

**Anything else we need to know?**:

Applying the second YAML and then manually creating the NSG in Azure does not actually allow traffic from Azure Front Door. I'm wondering why this might be...

**Environment**:
- Kubernetes version (use `kubectl version`):
```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.5", GitCommit:"20c265fef0741dd71a66480e35bd69f18351daea", GitTreeState:"clean", BuildDate:"2019-10-15T19:16:51Z", GoVersion:"go1.12.10", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.10", GitCommit:"89d8075525967c7a619641fabcb267358d28bf08", GitTreeState:"clean", BuildDate:"2020-06-23T02:52:37Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
```
- Size of cluster (how many worker nodes are in the cluster?)
`3 worker nodes`

- General description of workloads in the cluster (e.g. HTTP microservices, Java app, Ruby on Rails, machine learning, etc.)
HTTP(S) based micro-services using NGINX ingress

**Comment by nmiodice** [comment]
Also do you know when the patch will be rolled out? I did not see an ETA or version tag to look out for.

**Comment by nmiodice** [comment]
Also, my previous statement about configuring an internally routable IP address is incorrect.

My understanding of the problem has developed, and it would be helpful to get confirmation. It appears that the `loadBalancerSourceRanges` property is applied both at the NSG *and* within the cluster. This is why I needed to revert the change to `loadBalancerSourceRanges` in order for my manually configured NSG rule to allow traffic to the cluster.

In other words, the `loadBalancerSourceRanges` is applied both at the Azure NSG *and* on the cluster itself. Can you please confirm this?

**Comment by yangl900** [comment]
Hi @nmiodice , my understanding is that you don't need specify `loadBalancerSourceRanges`, since you are trying to use frontdoor service tag. 

The issue of the bug is that, it adds the 0.0.0.0/0 rule which makes the NSG useless. And if you specify `loadBalancerSourceRanges`, it will be used to construct the source IP prefixes of NSG rules here: https://github.com/kubernetes/kubernetes/blob/3229073042825666f4531787845cfc320d6ae786/staging/src/k8s.io/legacy-cloud-providers/azure/azure_loadbalancer.go#L1181

So unless you specify the IP ranges of Azure frontdoor in the `loadBalancerSourceRanges`, the traffic coming from it will be blocked, that is the bug.

**Comment by yangl900** [comment]
btw I don't think `loadBalancerSourceRanges` applies within the cluster, @feiskyer do you know?

**Comment by nmiodice** [comment]
Thanks @yangl900 . From my testing, it does appear that `loadBalancerSourceRanges` does apply within the cluster in some way... For example, if I set the `loadBalancerSourceRanges` to `127.0.0.1/32` and then delete the NSG rule that is created, the traffic is still blocked.

**Comment by yangl900** [comment]
you are right! it's indeed on iptable rules too: https://github.com/kubernetes/kubernetes/blob/2585d78c79054a6d85620eb5430a1498bb747967/pkg/proxy/iptables/proxier.go#L1184

**Comment by nmiodice** [comment]
Thanks! Being relatively new to Kubernetes, I was unsure where to look for the internal cluster configuration.

For now, I have configured `loadBalancerSourceRanges` to match the IP Ranges [published by Microsoft](https://www.microsoft.com/en-us/download/details.aspx?id=56519) but do not like this solution as these IP addresses are subject to change.

Is there any other workaround I should be aware of?

**Comment by yangl900** [comment]
@nmiodice I think that's the only option right now. But the fix should be available real soon too, the k8s patch version releases every month and should be available in AKS in 1-2 weeks once it's released in upstream project.

**Comment by yangl900** [comment]
Actually I think the upstream project already released and available. It's in v1.16.13 release and it's available in AKS. @nmiodice could you give a try?

the cherry pick merged on July 9th and released on July 15th: https://github.com/kubernetes/kubernetes/commits/release-1.16?after=4266f22fa4a0270bd2368ebafa36868ce44b6648+34&branch=release-1.16

**Comment by nmiodice** [comment]
@yangl900 , thank you so much for pointing this out. After upgrading to `1.16.13` the issue is resolved. Thanks!
